### ISSUE TITLE
[Inductor] Direct generalized_scatter copy-back changes returned output aliasing

### ISSUE THREAD

**Issue by ruddyscent** [pytorch/pytorch#195451]
I found a correctness issue in Inductor’s direct generalized-scatter copy-back path. The values are correct, but the aliasing behavior differs from eager mode: the compiled function returns the mutated input rather than a fresh tensor.

> **AI-assisted technical report**
>
> I reproduced a correctness issue in the direct generalized-scatter copy-back
> path. It preserves initial values but changes the returned tensor's alias
> contract: eager returns a fresh result, while Inductor returns the mutated
> input itself.
>
> ### Minimal reproducer
>
> ```python
> import torch
>
>
> def fn(x, src):
>     updated = torch.slice_scatter(x, src, 0, 0, 1)
>     x.copy_(updated)
>     return updated
>
>
> def run(callable):
>     x = torch.tensor([1.0, 2.0])
>     out = callable(x, torch.tensor([10.0]))
>     torch.testing.assert_close(x, torch.tensor([10.0, 2.0]))
>     torch.testing.assert_close(out, torch.tensor([10.0, 2.0]))
>     return x, out
>
>
> eager_x, eager_out = run(fn)
> print("eager aliases input:", torch._C._is_alias_of(eager_x, eager_out))
> eager_out.add_(100.0)
> print("eager input after output mutation:", eager_x)
>
> compiled_x, compiled_out = run(torch.compile(fn, fullgraph=True))
> print("compiled aliases input:", torch._C._is_alias_of(compiled_x, compiled_out))
> compiled_out.add_(100.0)
> print("compiled input after output mutation:", compiled_x)
> ```
>
> Expected eager behavior:
>
> ```text
> eager aliases input: False
> eager input after output mutation: tensor([10.,  2.])
> ```
>
> Actual compiled behavior:
>
> ```text
> compiled aliases input: True
> compiled input after output mutation: tensor([110., 102.])
> ```
>
> The same behavior reproduced with both `torch.compile(fn)` and
> `torch.compile(fn, fullgraph=True)` on CPU.
>
> ### Diagnosis and scope
>
> Functionalization creates this shape:
>
> ```text
> generalized_scatter(input, ...) -> copy_(input, scatter) -> return scatter
> ```
>
> `should_reinplace_scatter()` treats the direct copy-back as profitable and
> reinplaces the scatter. That is valid only when the functional scatter result
> is dead after copy-back. Here it is returned, so reinplacing changes a
> non-aliasing functional result into an alias of the input.
>
> A narrow direction is to apply a post-copy data-use/alias-liveness check to
> the direct copy-back case as well, including view and multi-output-view
> (`getitem`) aliases. The exact guard needs separate validation; this report
> does not propose an unconditional profitability change.
>
> ### Relation to existing work
>
> - #195285 and draft PR #195292 concern the indirect
>   `generalized_scatter -> index_put/_unsafe_index_put -> copy_(input, ...)`
>   traversal, motivated by a full-buffer clone. This report has no indexed
>   update and concerns return-alias correctness.
> - The fork reference implementation at
>   `6092aca85937d19257a8a1569c7d434b11313f1b` addresses liveness across the
>   indexed-update traversal, but intentionally leaves the pre-existing direct
>   copy-back fast path unchanged.
> - #195329 is a performance issue where the returned view intentionally aliases
>   the input. This reproducer is the opposite: eager returns an independent
>   result and compiled execution incorrectly creates an alias.
>
> ### Local verification
>
> Reproduced on CPU using the current Python compiler source checkout with an
> ABI-compatible local editable runtime. CUDA was unavailable in this
> environment.

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @bdhirsh @bobrenjc93 @aorenste

**Comment by HussainNizamani** [comment]
**Reproduced on two builds, and a defensive `.clone()` inside the compiled function doesn't survive either**

Confirmed on my end. I ran this on two separate builds on aarch64 CPU, a recent nightly wheel (`2.15.0.dev20260824+cpu`) and a local source build (`2.14.0a0+git9c98593`), and got identical results on both. 

Every version I tried triggers it, the MWE as written, plain `torch.compile(fn)` without fullgraph, and a realistic optimizer style pattern where an in place op applied to the returned tensor after the step silently changed the parameter as well.

I tried the obvious defensive fix, returning `updated.clone()` from inside the compiled function, and Inductor removes that clone too.

The output still aliases the input. 

Here are analysis from my AI:

>
> **What breaks.** In eager mode `fn` returns a tensor that is independent of `x`, so mutating the return value afterwards leaves `x` untouched. Under `torch.compile` the returned tensor is an alias of `x`, and any later in-place op on the output writes into the input. There is no error or warning at any point. A concrete consequence, using an optimizer-style caller that applies `new_param.mul_(0.9)` after the step: eager keeps `param = [10., 2.]`, compiled ends with `param = [9.0, 1.8]`.
>
> **Root cause.** Consistent with the diagnosis in the issue body. Functionalization produces `scatter -> copy_(input, scatter) -> return scatter`, and `should_reinplace_scatter()` treats the direct copy-back as profitable and reinplaces the scatter onto the input buffer without checking that the functional scatter result is still live as a graph output. Running the same MWE with `backend="aot_eager"` and `backend="eager"` gives correct, non-aliasing results, which localizes the bug to Inductor's reinplacing pass rather than functionalization or AOTAutograd. The clone experiment below suggests the surrounding buffer-reuse/copy-elimination then also removes a defensive clone whose source was reinplaced, though we did not trace that second interaction to a specific line.
>
> **Workarounds tested** (nightly wheel `2.15.0.dev20260824+cpu`, aarch64 CPU; each probe checks `torch._C._is_alias_of(x, out)` and whether mutating `out` corrupts `x`):
>
> | Probe | Output aliases input | Input corrupted |
> |---|---|---|
> | eager baseline | no | no |
> | `torch.compile` (Inductor, default) | **yes** | **yes** |
> | `torch.compile(fn)` without fullgraph | **yes** | **yes** |
> | `backend="aot_eager"` | no | no |
> | `backend="eager"` | no | no |
> | `return updated.clone()` inside the compiled function (Inductor) | **yes** | **yes** |
> | `compiled_fn(...)` then `.clone()` outside the compiled region | no | no |
>
> We also searched `torch._inductor.config` for a toggle covering this pass and found no config attribute matching `*inplace*` on this build, so there is no kill-switch style mitigation.
>
> **Why a workaround is not sufficient.** The only mitigation that works is cloning the result outside the compiled region, which requires every caller to already know about the bug. Code inside a compiled region cannot defend itself, since an explicit clone of the scatter result is eliminated. The failure is silent, changes the eager/compile aliasing contract, and shows up in ordinary patterns (update a buffer, return the result, mutate it later). That combination is what the `module: correctness (silent)` label exists for, and it argues for a liveness/alias check on the direct copy-back path in `should_reinplace_scatter()` rather than documentation or a user-side recipe.

**Comment by wwoosshh** [comment]
The same path also produces wrong *values* when the later mutation of the input is unrelated to the scatter result:

    def fn(x, src):
        y = torch.slice_scatter(x, src[1:3], dim=0, start=1, end=3)
        x.zero_()
        return y      # eager: the scattered values; inductor: all zeros

(`select_scatter` + `fill_` behaves the same; 2.14.0 and 2.15.0.dev20260921, CUDA.) can_inplace only
checks that *some* copy_ into the input exists, not that it writes this node's result back, and does not
check whether the result is observed after that copy_. PR: #198096 (issue #198094)
